### PR TITLE
[debug dump util] Base Skeleton and CLI top level class added

### PR DESCRIPTION
**Pull request by vivekreddynv**
Signed-off-by: Vivek Reddy Karri <vkarri@nvidia.com>

#### What I did
HLD for Dump Utility: [HLD](https://github.com/vivekreddynv/SONiC/pull/2).
* Added the top level CLI command i.e "dump state"
* Added corresponding UT's
* Added the autocompletion support
* Added the implementation for the customization options provided, such as --db, --key-map, --show, --table & --namespace

#### How to verify it

UT's are implemented

```
tests/dump_tests/dump_state_test.py::TestDumpState::test_identifier_single PASSED                                                       [  8%]
tests/dump_tests/dump_state_test.py::TestDumpState::test_identifier_multiple PASSED                                                     [ 16%]
tests/dump_tests/dump_state_test.py::TestDumpState::test_option_key_map PASSED                                                          [ 25%]
tests/dump_tests/dump_state_test.py::TestDumpState::test_option_db_filtering PASSED                                                     [ 33%]
tests/dump_tests/dump_state_test.py::TestDumpState::test_option_tabular_display PASSED                                                  [ 41%]
tests/dump_tests/dump_state_test.py::TestDumpState::test_option_tabular_display_no_db_filter PASSED                                     [ 50%]
tests/dump_tests/dump_state_test.py::TestDumpState::test_identifier_all_with_filtering PASSED                                           [ 58%]
tests/dump_tests/dump_state_test.py::TestDumpState::test_namespace_single_asic PASSED                                                   [ 66%]
tests/dump_tests/dump_state_test.py::TestDumpStateMultiAsic::test_default_namespace PASSED                                              [ 75%]
tests/dump_tests/dump_state_test.py::TestDumpStateMultiAsic::test_namespace_asic0 PASSED                                                [ 83%]
tests/dump_tests/dump_state_test.py::TestDumpStateMultiAsic::test_namespace_asic1 PASSED                                                [ 91%]
tests/dump_tests/dump_state_test.py::TestDumpStateMultiAsic::test_invalid_namespace PASSED                                              [100%]
```

```
admin@sonic:~$ dump state --help
Usage: dump state [OPTIONS] MODULE IDENTIFIER

  Dump the redis-state of the identifier for the module specified

Options:
  -s, --show            Display Modules Available
  -d, --db TEXT         Only dump from these Databases
  -t, --table           Print in tabular format  [default: False]
  -k, --key-map         Only fetch the keys matched, don't extract field-value
                        dumps  [default: False]
  -v, --verbose         Prints any intermediate output to stdout useful for
                        dev & troubleshooting  [default: False]
  -n, --namespace TEXT  Dump the redis-state for this namespace.  [default: ]
  --no-split            Doesn't split the identifier when a list type (eg:
                        1,2,3,4) is passed as an identifier  [default: False]
  --help                Show this message and exit.


admin@sonic:~$ dump state --show
Module    Identifier
--------  ------------
copp      trap_id
port      port_name

admin@sonic:~$ dump state port Ethernet0 --table --db ASIC_DB --key-map
+-------------+-----------+-----------------------------------------------------------+
| port_name   | DB_NAME   | DUMP                                                      |
+=============+===========+===========================================================+
| Ethernet0   | ASIC_DB   | +-------------------------------------------------------+ |
|             |           | | Keys Collected                                        | |
|             |           | +=======================================================+ |
|             |           | | ASIC_STATE:SAI_OBJECT_TYPE_HOSTIF:oid:0xd000000000a21 | |
|             |           | +-------------------------------------------------------+ |
|             |           | | ASIC_STATE:SAI_OBJECT_TYPE_PORT:oid:0x10000000008d4   | |
|             |           | +-------------------------------------------------------+ |
|             |           | +---------------------+---------------------+             |
|             |           | | vid                 | rid                 |             |
|             |           | +=====================+=====================+             |
|             |           | | oid:0xd000000000a21 | oid:0xd             |             |
|             |           | +---------------------+---------------------+             |
|             |           | | oid:0x10000000008d4 | oid:0x1690000000001 |             |
|             |           | +---------------------+---------------------+             |
+-------------+-----------+-----------------------------------------------------------+

admin@sonic:~$ dump state port Ethernet0 --db ASIC_DB --key-map
{
    "Ethernet0": {
        "ASIC_DB": {
            "keys": [
                "ASIC_STATE:SAI_OBJECT_TYPE_HOSTIF:oid:0xd000000000a21",
                "ASIC_STATE:SAI_OBJECT_TYPE_PORT:oid:0x10000000008d4"
            ],
            "tables_not_found": [],
            "vidtorid": {
                "oid:0xd000000000a21": "oid:0xd",
                "oid:0x10000000008d4": "oid:0x1690000000001"
            }
        }
    }
}
```

Examples Output samples can be found in the HLD  [Link](https://github.com/vivekreddynv/SONiC/blob/debug_dump_utility/doc/Dump-Utility.md#12-customization-options)

### CODE REVIEW

File: dump/main.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    if len(db) > 0:
+        collected_info = filter_out_dbs(db, collected_info)
+    
+    vidtorid = extract_rid(collected_info, namespace)
+    
+    if not key_map:
+        collected_info = populate_fv(collected_info, module, namespace)
+    
+    for id in vidtorid.keys():
+        if  vidtorid[id] and vidtorid:
```
Comment (dgsudharsan) [inline dump/main.py:95]: Please remove the second condition after and

File: dump/main.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+         
+    print_dump(collected_info, table, module, identifier, key_map)
+    
+    return 
+
+def extract_rid(info, ns):
+    r = SonicV2Connector(namespace=ns, host="127.0.0.1")
+    r.connect("ASIC_DB")
+    vidtorid = {}
+    for arg in info.keys():
```
Comment (dgsudharsan) [inline dump/main.py:106]: Can you make sure you fill the arg only when the get_v_r_map returns a non empty object? In this way we can avoid the if condition in the for loop above

File: dump/main.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+def get_keys(dump):
+    keys = []
+    for key_ in dump:
+        if isinstance(key_, dict) and key_:
+            keys.append(list(key_.keys())[0])
+        else:
+            keys.append(key_)
+    return keys
+
+def get_rid(redis_key, vidtorid):
```
Comment (dgsudharsan) [inline dump/main.py:183]: Please remove this function if it is not invoked anywhere

File: dump/main.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+                total_info += str(tabulate(values, hdrs, tablefmt="grid"))
+            else:
+                temp = []
+                for key_ in collected_info[ids][db]["keys"]:
+                    temp.append([key_])
+                total_info += str(tabulate(temp, headers=["Keys Collected"], tablefmt="grid"))
+
+            total_info += "\n"
+            if "vidtorid" in collected_info[ids][db]:
+                table = []
```
Comment (dgsudharsan) [inline dump/main.py:223]: The name for this variable is same as one of the arguments to the API. Please change.

**Review by dgsudharsan — CHANGES_REQUESTED** [review]

File: dump/main.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    asic_obj_ptrn = "ASIC_STATE:.*:oid:0x\w{1,14}"
+    
+    if "ASIC_DB" in single_dict and 'keys' in single_dict["ASIC_DB"]:
+        for redis_key in single_dict["ASIC_DB"]['keys']:
+            if re.match(asic_obj_ptrn, redis_key):
+                matches = re.findall(r"oid:0x\w{1,14}", redis_key)
+                if matches:
+                   vid = matches[0]
+                   v_r_map[vid] =  vid_to_rid(vid, r)
+    return v_r_map
```
Comment (stepanblyschak) [inline dump/main.py:113]: This function returns VID->RID mapping? Can you just return VIDTORID content from ASIC_DB?

File: dump/main.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+# Get a vid:rid for the input vid
+def vid_to_rid(vid, r):
+    rid = r.get("ASIC_DB", "VIDTORID", vid)
+    if not rid:
+        rid = "Real ID Not Found" 
+    return rid  
+
+# Filter dbs which are not required
+def filter_out_dbs(db_list, collected_info):
+    args_ = list(collected_info.keys())
```
Comment (stepanblyschak) [inline dump/main.py:124]: Why create args_ ? list convertion is not required here.

File: dump/main.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    for arg in args_:
+        dbs = list(collected_info[arg].keys())
```
Comment (stepanblyschak) [inline dump/main.py:126]: list convertion is not required here

File: tests/dump_tests/dump_state_test.py [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+        # and therefore ignoring these paths. field-value dump capability of the utility is nevertheless verified using f-v dumps of ASIC_DB & STATE_DB
+        pths = ["root['Ethernet0']['CONFIG_DB']['keys'][0]['PORT|Ethernet0']", "root['Ethernet0']['APPL_DB']['keys'][0]['PORT_TABLE:Ethernet0']"]
+        ddiff = compare_json_output(expected, result.output, exclude_paths = pths)
+        assert not ddiff, ddiff
+        
+    def test_identifier_multiple(self):
+        runner = CliRunner()
+        result = runner.invoke(dump.state, ["port", "Ethernet0,Ethernet4"])
+        print(result.output)
+        expected = ('''{"Ethernet0":{"CONFIG_DB":{"keys":[{"PORT|Ethernet0":{"alias":"etp1","description":"etp1","index":"0","lanes":"25,26,27,28",''' +
```
Comment (stepanblyschak) [inline tests/dump_tests/dump_state_test.py:106]: Can this be an actual dict not a string allowing easier modification in future

**Review by stepanblyschak — CHANGES_REQUESTED** [review]

File: dump/main.py (hunk 7 above)
Comment (vivekreddynv) [reply dump/main.py:126]: same applies here

File: tests/dump_tests/dump_state_test.py (hunk 8 above)
Comment (vivekreddynv) [reply tests/dump_tests/dump_state_test.py]: Done

File: dump/main.py (hunk 6 above)
Comment (stepanblyschak) [reply dump/main.py:119]: collected_info.keys() returns an iterable, you can loop over it

**Review by dgsudharsan — APPROVED** [review]

**Pull request closed without merging**